File: Protobuffprotocol/ProtoUtils.cs
using System;
using System.Collections;
using System.Collections.Generic;
using LabMPP.domain;

using Protobuffprotocol;
using FestivalDTO = LabMPP.domain.FestivalDTO;
using proto=Protobuffprotocol;
namespace protobuf
{
    public class ProtoUtils
    {
        public static String getError( proto.Response response) {
        String errorMessage = response.Error;
        return errorMessage;
    }

    public static Request createLoginRequest(LabMPP.domain.Account user)
    {
        proto.Account acc = new proto.Account {Username = user.id,Password = user.password,Name = user.name};
        Request request = new proto.Request {Type = proto.Request.Types.Type.Login, User = acc};
        return request;
    }

    public static  proto.Request createLogoutRequest(LabMPP.domain.Account user) {
        proto.Account acc = new proto.Account {Username = user.id,Password = user.password,Name = user.name};
        Request request = new proto.Request {Type = proto.Request.Types.Type.Login, User = acc};
        return request;
    }

    public static LabMPP.domain.Account getUser( proto.Response response) {
        LabMPP.domain.Account user=new LabMPP.domain.Account(response.User.Username, response.User.Password, response.User.Name);
        return user;
    }

    public static LabMPP.domain.Account getUser( proto.Request request){
        LabMPP.domain.Account user=new LabMPP.domain.Account(request.User.Username, request.User.Password, request.User.Name);
        return user;
    }
    public static  proto.Response createOkResponse()
    {
        proto.Response response = new proto.Response {Type = proto.Response.Types.Type.Ok};
        return response;
    }

    public static  proto.Response createErrorResponse(String message) {
        proto.Response response= new proto.Response {Type = proto.Response.Types.Type.Error};
        return response;
    }

    public static  proto.Response createLoginResponse(LabMPP.domain.Account account) {
         proto.Account user= new proto.Account {Username = account.id,Password = account.password,Name = account.name};
         proto.Response response=   new proto.Response {Type = proto.Response.Types.Type.Login, User = user};
         
        return response;
    }

    public static IEnumerable<LabMPP.domain.FestivalDTO> getFestivals( proto.Response response) {
        LabMPP.domain.FestivalDTO[] festivalDTOS = new LabMPP.domain.FestivalDTO[response.Festivals.Count];
        for(int i=0;i<response.Festivals.Count;i++)
        {
            proto.FestivalDTO festivalDTO = response.Festivals[i];
             LabMPP.domain.FestivalDTO festival = new LabMPP.domain.FestivalDTO(festivalDTO.Name, DateTime.Parse(festivalDTO.Date),TimeSpan.Zero, festivalDTO.Location, (int)festivalDTO.Seats,(int)festivalDTO.SoldSeats, (int)festivalDTO.FestivalID);
            festivalDTOS[i]=festival;
        }

        return festivalDTOS;
    }

    public static  proto.Request createFestivalRequest(DateTime date) {
         proto.Request request= new proto.Request {Type = proto.Request.Types.Type.SearchByDate, Date = date.ToString("yyyy-MM-dd")};  
        return request;
    }
    public static  proto.Request createFestivalRequest() {
         proto.Request request=  new proto.Request {Type = proto.Request.Types.Type.SearchByDate};  
        return request;
    }

    public static  proto.Response createFestivalResponse(IEnumerable<FestivalDTO> festivalDTOS) {
         proto.Response response=  new proto.Response {Type = proto.Response.Types.Type.SearchByDate};
         foreach (var festivalDTO in festivalDTOS)
         {
             proto.FestivalDTO festival = new proto.FestivalDTO
             {
                 FestivalID = festivalDTO.Id, Date = festivalDTO.Data.ToString("yyyy-MM-dd"), Name = festivalDTO.Name,
                 Location = festivalDTO.Location, Seats = festivalDTO.AvailableSeats, SoldSeats = festivalDTO.SoldSeats
             };
             response.Festivals.Add(festival);
         }

         return response;
    }

    public static  proto.Request createSellTicketRequest(LabMPP.domain.TicketDTO ticketDTO)
    {
        proto.TicketDTO ticket = new proto.TicketDTO
            {Client = ticketDTO.client, Seats = ticketDTO.seats, FestivalID = ticketDTO.festivalID};
        proto.Request request = new Request {Type = Request.Types.Type.SellTicket, Ticket = ticket};
        return request;
    }

    public static LabMPP.domain.TicketDTO getTicket( proto.TicketDTO ticket) {
        LabMPP.domain.TicketDTO ticketDTO = new LabMPP.domain.TicketDTO(ticket.FestivalID,ticket.Seats,ticket.Client);
        return  ticketDTO;
    }

    public static  proto.Response createTicketSoldResponse(LabMPP.domain.TicketDTO ticket) {
        proto.TicketDTO ticketDTO = new proto.TicketDTO
            {Client = ticket.client, Seats = ticket.seats, FestivalID = ticket.festivalID};
         proto.Response response= new Response {Type = Response.Types.Type.SellTicket, Ticket = ticketDTO};
        return response;
    }
    }
}

File: NetworkingProtobuff/src/main/java/protobuffprotocol/ProtoUtils.java
package protobuffprotocol;


import domain.Account;
import domain.FestivalDTO;
import domain.TicketDTO;

import java.sql.Date;
import java.util.Arrays;

public class ProtoUtils {
    public static String getError(FestivalProtocol.Response response) {
        String errorMessage = response.getError();
        return errorMessage;
    }

    public static FestivalProtocol.Request createLoginRequest(Account user) {
        FestivalProtocol.Account acc = FestivalProtocol.Account.newBuilder().setUsername(user.getUsername()).setPassword(user.getPassword()).build();
        FestivalProtocol.Request request= FestivalProtocol.Request.newBuilder().setType(FestivalProtocol.Request.Type.Login)
                .setUser(acc).build();
        return request;
    }

    public static FestivalProtocol.Request createLogoutRequest(Account user) {
        FestivalProtocol.Account acc = FestivalProtocol.Account.newBuilder().setUsername(user.getUsername()).setPassword(user.getPassword()).build();
        FestivalProtocol.Request request= FestivalProtocol.Request.newBuilder().setType(FestivalProtocol.Request.Type.Logout)
                .setUser(acc).build();
        return request;
    }

    public static Account getUser(FestivalProtocol.Response response) {
        Account user=new Account(response.getUser().getUsername(), response.getUser().getPassword(), response.getUser().getName());
        return user;
    }

    public static Account getUser(FestivalProtocol.Request request){
        Account user=new Account(request.getUser().getUsername(), request.getUser().getPassword(), request.getUser().getName());
        return user;
    }
    public static FestivalProtocol.Response createOkResponse() {
        FestivalProtocol.Response response=FestivalProtocol.Response.newBuilder()
                .setType(FestivalProtocol.Response.Type.Ok).build();
        return response;
    }

    public static FestivalProtocol.Response createErrorResponse(String message) {
        FestivalProtocol.Response response=FestivalProtocol.Response.newBuilder()
                .setType(FestivalProtocol.Response.Type.Error)
                .setError(message).build();
        return response;
    }

    public static FestivalProtocol.Response createLoginResponse(Account account) {
        FestivalProtocol.Account user=FestivalProtocol.Account.newBuilder()
                .setUsername(account.getUsername()).setName(account.getName()).setPassword(account.getPassword()).build();
        FestivalProtocol.Response response=  FestivalProtocol.Response.newBuilder()
                .setType( FestivalProtocol.Response.Type.Login)
                .setUser(user).build();
        return response;
    }

    public static Iterable<FestivalDTO> getFestivals(FestivalProtocol.Response response) {
        FestivalDTO[] festivalDTOS = new FestivalDTO[response.getFestivalsCount()];
        for(int i=0;i<response.getFestivalsCount();i++){
            FestivalProtocol.FestivalDTO festivalDTO=response.getFestivals(i);
            FestivalDTO festival = new FestivalDTO(festivalDTO.getFestivalID(),festivalDTO.getName(), Date.valueOf(festivalDTO.getDate().split(" ")[0]),festivalDTO.getLocation(), festivalDTO.getSeats(),festivalDTO.getSoldSeats());
            festivalDTOS[i]=festival;
        }
        return Arrays.asList(festivalDTOS.clone());
    }

    public static FestivalProtocol.Request createFestivalRequest(Date date) {
        FestivalProtocol.Request request= FestivalProtocol.Request.newBuilder().setType(FestivalProtocol.Request.Type.SearchByDate)
                .setDate(date.toString()).build();
        return request;
    }
    public static FestivalProtocol.Request createFestivalRequest() {
        FestivalProtocol.Request request= FestivalProtocol.Request.newBuilder().setType(FestivalProtocol.Request.Type.SearchByDate)
                .build();
        return request;
    }

    public static FestivalProtocol.Response createFestivalResponse(Iterable<FestivalDTO> festivalDTOS) {
        FestivalProtocol.Response.Builder response= FestivalProtocol.Response.newBuilder()
                .setType(FestivalProtocol.Response.Type.SearchByDate);
        for (FestivalDTO festivalDTO: festivalDTOS){
            FestivalProtocol.FestivalDTO festival=FestivalProtocol.FestivalDTO.newBuilder()
                    .setFestivalID(festivalDTO.getFestivalID()).setDate(festivalDTO.getDate().toString())
                    .setLocation(festivalDTO.getLocation()).setName(festivalDTO.getName())
                    .setSeats(festivalDTO.getSeats()).setSoldSeats(festivalDTO.getSoldSeats()).build();
            response.addFestivals(festival);
        }

        return response.build();
    }

    public static FestivalProtocol.Request createSellTicketRequest(TicketDTO ticketDTO) {
        FestivalProtocol.TicketDTO ticket = FestivalProtocol.TicketDTO.newBuilder().setFestivalID(ticketDTO.getFestivalID())
                .setClient(ticketDTO.getClient()).setSeats(ticketDTO.getSeats()).build();
        FestivalProtocol.Request request= FestivalProtocol.Request.newBuilder().setType(FestivalProtocol.Request.Type.SellTicket)
                .setTicket(ticket).build();
        return request;
    }

    public static TicketDTO getTicket(FestivalProtocol.TicketDTO ticket) {
        TicketDTO ticketDTO = new TicketDTO(ticket.getFestivalID(),ticket.getSeats(),ticket.getClient());
        return  ticketDTO;
    }

    public static FestivalProtocol.Response createTicketSoldResponse(TicketDTO ticket) {
        FestivalProtocol.TicketDTO ticketDTO = FestivalProtocol.TicketDTO.newBuilder().setFestivalID(ticket.getFestivalID())
                .setClient(ticket.getClient()).setSeats(ticket.getSeats()).build();
        FestivalProtocol.Response response=FestivalProtocol.Response.newBuilder()
                .setType(FestivalProtocol.Response.Type.SellTicket).setTicket(ticketDTO).build();
        return response;
    }
}
